lib/index.ts

```diff
@@ -21,13 +41,323 @@ export function normalizeTarEntry(name: string): string {
+	});
+}
+
+/**
+ * Return a promise that resolves to a new tar.Pack stream that duplicates the
+ * contents of the given sourceTarStream. The promise resolves when the source
+ * stream is fully consumed. This is useful to trigger stream events on
+ * listeners attached in advance by the caller to sourceTarStream, which may
+ * itself be a tar.Pack stream (though any Readable will do). The new Pack
+ * stream is fully cached in memory, so watch out for memory usage.
+ * @param sourceTarStream Source tar stream, possibly a tar.Pack object
+ * @param opts.onEntry Optional callback called for every entry in the tar
+ *     stream. onEntry may return a promise which will then be awaited for.
+ * @param opts.onFinish Callback called right before pack.finalize(), allowing
+ *     the caller to insert additional tar entries. onFinish may return a
+ *     promise which will then be awaited for.
+ */
+export function cloneTarStream(
+	sourceTarStream: Readable,
+	opts?: {
+		onEntry?: (pack: tar.Pack, header: tar.Headers, stream: Readable) => void;
+		onFinish?: (pack: tar.Pack) => void;
+	},
+): Bluebird<tar.Pack> {
+	const extract = tar.extract();
+	const pack = tar.pack();
+	let packOnError: (error: Error) => void;
+	let sourceTarStreamOnError: (error: Error) => void;
+	return new Bluebird<tar.Pack>((resolve, reject) => {
+		packOnError = reject;
+		sourceTarStreamOnError = reject;
+		sourceTarStream.on('error', sourceTarStreamOnError);
+		pack.on('error', packOnError);
+		extract.on('error', reject);
+		extract.on(
+			'entry',
+			(header: tar.Headers, stream: Readable, callback: tar.Callback) => {
+				if (opts && opts.onEntry) {
+					Bluebird.try(() => opts.onEntry!(pack, header, stream)).then(
+						callback as () => void,
+						callback,
+					);
+				} else {
+					streamToBuffer(stream).then(buf => {
+						pack.entry(header, buf, callback);
+					});
+				}
+			},
+		);
+		extract.once('finish', () => {
+			Bluebird.try(() => {
+				if (opts && opts.onFinish) {
+					return opts.onFinish(pack);
+				}
+			})
+				.then(() => {
+					pack.finalize();
+					resolve(pack);
+				})
+				.catch(reject);
+		});
+		sourceTarStream.pipe(extract);
+	}).finally(() => {
+		sourceTarStream.removeListener('error', sourceTarStreamOnError || noop);
+		pack.removeListener('error', packOnError || noop);
+	});
+}
+
+/**
+ * Read fromStream in small chunks, and write each chunk to each stream in
+ * the given toStreams array. This is a memory-efficient way of "multicasting"
+ * the contents of fromStream to each of the destination toStreams, while
+ * respecting their individual flow-control mechanism ('drain' events).
+ * Return a promise that resolves when the operation "fully completes",
+ * i.e. when the 'finish' event has been emitted for every stream in the
+ * toStreams array.
+ * @param fromStream
+ * @param toStreams
+ */
+export function multicastStream(
+	fromStream: Readable,
+	toStreams: Writable[],
+): Bluebird<void> {
```

Comment:
Conceptually, what `multicastStream` does is simple: it pipes a single source stream to a collection of destination streams provided by the caller, in small chunks and thus in a memory efficient way. It "listens" to the drain events of each destination stream separately, thus also works for mixed destination streams that consume the data at different rates. If any one destination applies backpressure, piping is paused for all destinations (to keep memory usage low). The implementation turned out to be more code than I expected, but hopefully not too hard to understand. I have written a few tests for it, including the case of destination streams that consume data at different rates, so I'm fairly confident it works. :-)
Sadly, the use case I wrote it for turned out not to be feasible. I had hoped to use it to replace some whole-file buffering in `resin-multibuild` ([here](https://github.com/balena-io-modules/resin-multibuild/blob/v1.0.0/lib/index.ts#L83-L90)), but then I learned (the hard way) that `tar-stream` treats `pack.entry(header, buf)` quite differently from `stream.pipe(pack.entry(header))`, enforcing stream backpressure only the to the latter while allowing the former to buffer as much memory as the system allows. Still, I think the `multicastStream` function is generic enough to have it as part of the toolkit (node-tar-utils).

Reply:
You can pipe a readable stream to many writable streams just using `readable.pipe(writableX)` multiple times.

Reply:
@zvin, I should stop adding you as reviewer to my pull requests! :-)
I was misled by this StackOverflow post where people seem to conclude that running `readable.pipe(writableX)` does not work: https://stackoverflow.com/questions/19553837/node-js-piping-the-same-readable-stream-into-multiple-writable-targets
But I suppose their circumstances are different. I have now pushed a much shorter implementation doing just `pipe` multiple times as suggested. My awesome tests have all passed, so I assume it just works. :-)